File: packages/plugin-sdk/opencapx_sdk/protocol.py
"""JSON-RPC 2.0 over stdio framing for OpenCapX plugins.

One JSON object per line, UTF-8, no BOM, `\\n`-separated.
stdout carries only protocol; stderr is left for logs.

See docs/plugin-protocol.md.
"""

from __future__ import annotations

import json
import sys
from typing import Any, Callable, Iterable, Iterator, Optional, TextIO


class ProtocolError(Exception):
    """Raised when a JSON-RPC frame/field is invalid."""


def write_message(msg: dict[str, Any], out: Optional[TextIO] = None) -> None:
    """Serialize + write one line of JSON to stdout (default) or the given stream."""
    stream = out if out is not None else sys.stdout
    stream.write(json.dumps(msg, ensure_ascii=False) + "\n")
    stream.flush()


def read_message(
    lines: Optional[Iterable[str]] = None,
    line_factory: Optional[Callable[[], str]] = None,
) -> Optional[dict[str, Any]]:
    """Read and parse one line of a JSON-RPC message. Returning None means EOF.

    lines: an iterable of strings (defaults to sys.stdin).
    line_factory: a callable that returns one line per call. If neither is passed, read line by line from sys.stdin.

    EOF semantics:
      - From sys.stdin: `readline()` returning "" = EOF; a line that strips to empty = skipped.
      - From iterating lines: StopIteration = EOF; an empty-string element = skipped.
      - line_factory: decided by the caller; returning "" is always treated as EOF (simple behavior).
    """
    use_iter: Optional[Iterator[str]] = None
    if line_factory is not None:
        getter: Callable[[], str] = line_factory
        stdin_mode = False
    elif lines is not None:
        use_iter = iter(lines)
        stdin_mode = False
    else:
        stdin_mode = True

    while True:
        if use_iter is not None:
            try:
                raw = next(use_iter)
            except StopIteration:
                return None
            stripped = raw.strip()
            if not stripped:
                continue  # skip empty/whitespace-only lines
            try:
                return json.loads(stripped)
            except json.JSONDecodeError as e:
                raise ProtocolError(f"malformed json: {e.msg} (line: {stripped[:80]!r})") from e
        elif stdin_mode:
            raw = sys.stdin.readline()
            if not raw:
                return None  # sys.stdin EOF
            stripped = raw.strip()
            if not stripped:
                continue
            try:
                return json.loads(stripped)
            except json.JSONDecodeError as e:
                raise ProtocolError(f"malformed json: {e.msg} (line: {stripped[:80]!r})") from e
        else:
            raw = getter()
            if not raw:
                return None
            stripped = raw.strip()
            if not stripped:
                continue
            try:
                return json.loads(stripped)
            except json.JSONDecodeError as e:
                raise ProtocolError(f"malformed json: {e.msg} (line: {stripped[:80]!r})") from e


def build_result(req_id: Any, result: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": req_id, "result": result}


def build_error(req_id: Any, code: int, message: str, data: Any = None) -> dict[str, Any]:
    err: dict[str, Any] = {"code": code, "message": message}
    if data is not None:
        err["data"] = data
    return {"jsonrpc": "2.0", "id": req_id, "error": err}


def build_request(method: str, params: Any, req_id: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": req_id, "method": method, "params": params}


def build_notification(method: str, params: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "method": method, "params": params}

File: packages/ts-sdk/src/protocol.ts
/** JSON-RPC 2.0 over stdio framing for OpenCapX plugins.
 *
 * One JSON object per line, UTF-8, `\n` separated.
 * stdout carries protocol only; stderr is for logs.
 *
 * Mirrors `packages/plugin-sdk/opencapx_sdk/protocol.py`.
 * See `docs/plugin-protocol.md`.
 */

/** JSON-RPC standard error codes. */
export const ERR_PARSE = -32700;
export const ERR_INVALID_REQUEST = -32600;
export const ERR_METHOD_NOT_FOUND = -32601;
export const ERR_INVALID_PARAMS = -32602;
export const ERR_INTERNAL = -32603;

/** Thrown when a JSON-RPC frame/field is illegal. */
export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolError";
  }
}

/** Any JSON value. */
export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

/** A decoded JSON-RPC message envelope. */
export interface JsonRpcMessage {
  jsonrpc?: unknown;
  id?: string | number | null;
  method?: unknown;
  params?: unknown;
  result?: unknown;
  error?: unknown;
}

export type JsonRpcRequestId = string | number;

/** Serialize + write one JSON line to stdout (default) or the given stream. */
export function writeMessage(
  msg: unknown,
  out?: NodeJS.WritableStream,
): void {
  const stream: NodeJS.WritableStream = out ?? process.stdout;
  stream.write(`${JSON.stringify(msg)}\n`);
}

/**
 * Decode a single input line. Returns `null` for blank/whitespace-only
 * lines (skip). Throws `ProtocolError` on malformed JSON or non-object
 * payloads.
 */
export function decodeLine(line: string): JsonRpcMessage | null {
  const stripped = line.trim();
  if (!stripped) return null;
  let parsed: unknown;
  try {
    parsed = JSON.parse(stripped);
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err);
    throw new ProtocolError(
      `malformed json: ${reason} (line: ${JSON.stringify(stripped.slice(0, 80))})`,
    );
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new ProtocolError(
      `malformed json: top-level must be an object (line: ${JSON.stringify(stripped.slice(0, 80))})`,
    );
  }
  return parsed as JsonRpcMessage;
}

export function buildResult(
  reqId: JsonRpcRequestId,
  result: unknown,
): Record<string, unknown> {
  return { jsonrpc: "2.0", id: reqId, result };
}

export function buildError(
  reqId: JsonRpcRequestId,
  code: number,
  message: string,
  data?: unknown,
): Record<string, unknown> {
  const err: Record<string, unknown> = { code, message };
  if (data !== undefined) err["data"] = data;
  return { jsonrpc: "2.0", id: reqId, error: err };
}

export function buildRequest(
  method: string,
  params: unknown,
  reqId: JsonRpcRequestId,
): Record<string, unknown> {
  return { jsonrpc: "2.0", id: reqId, method, params };
}

export function buildNotification(
  method: string,
  params: unknown,
): Record<string, unknown> {
  return { jsonrpc: "2.0", method, params };
}
